src/metaswitch/clearwater/queue_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,253 @@
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+import json
+import constants
+from time import sleep
+from concurrent import futures
+from metaswitch.clearwater.etcd_shared.common_etcd_synchronizer import CommonEtcdSynchronizer
+from .queue_fsm import QueueFSM
+import logging
+from etcd import EtcdAlreadyExist
+from queue_utils import remove_id_from_json, add_id_to_json, status_in_queue
+
+_log = logging.getLogger(__name__)
+
+class EtcdSynchronizer(CommonEtcdSynchronizer):
+    def __init__(self, plugin, ip, site, key, node_type, etcd_ip=None):
+        super(EtcdSynchronizer, self).__init__(plugin, ip, etcd_ip)
+        self.WAIT_FOR_TIMER_POP = 1
+        self._id = ip + "-" + node_type
+        self._timer_flag = False
+        self._fsm = QueueFSM(self._plugin, self._id, self.fsm_timer_expired)
+        self._site = site
+        self._key = key
+
+    def key(self):
+        return "/" + self._key + "/" + self._site + "/configuration/" + self._plugin.key()
+
+    def is_running(self):
+        return self._fsm.is_running()
+
+    def default_value(self):
+        return "{\"FORCE\": false, \"ERRORED\": [], \"COMPLETED\": [], \"QUEUED\": []}"
+ 
+    def main(self):
+        # Continue looping while the FSM is running.
+        while self._fsm.is_running():
+            _log.debug("Waiting for queue change from etcd")
+
+            etcd_result = None
+            self._abort_read = False
+
+            executor = futures.ThreadPoolExecutor(2)
+            fsm_timer_future = executor.submit(self.wait_for_fsm)
+            etcd_future = executor.submit(self.update_from_etcd)
+            futures.wait([etcd_future, fsm_timer_future], return_when=futures.FIRST_COMPLETED)
+
+            if self._terminate_flag:
+                executor.shutdown(wait=False)    
+                break
+
+            if fsm_timer_future.done():
```

Comment:
I feel like this needs a comment to explain what real-world situation it's handling.